src/bam_to_bed/intron_retention.py

```diff
@@ -0,0 +1,163 @@
+#    head -n 1 Hs_Ensembl_exon.txt
+#    gene	exon-id	chromosome	strand	exon-region-start(s)	exon-region-stop(s)	constitutive_call	ens_exon_ids	splice_events	splice_junctions
+#
+# 2. Convert to BED format excluding all not intron records
+#    cat Hs_Ensembl_exon.txt | grep -v "gene" | awk '$2 ~ /^I/ {print $3"\t"$5"\t"$6"\t"$2"-"$1"\t"0"\t"$4}' | sort -k1,1 -k2,2n -k3,3n | bgzip > hs_ref.bed.gz
+#    tabix -p bed hs_ref.bed.gz
+###############################################################################################################################################################
+
+
+#################################################################################################
+# DEPRECATED INSTRUCTIONS
+# keep it in case we decide to support GTF input in future
+#
+# (grep ^"#" input.gtf; grep -v ^"#" input.gtf | sort -k1,1 -k4,4n) | bgzip > input.sorted.gtf.gz
+# tabix -p gff input.sorted.gtf.gz
+#################################################################################################
+
+
+def is_paired(args):
+    with pysam.AlignmentFile(args.bam, mode="rb", threads=args.threads) as bam_handler:
+        bam_iter = bam_handler.fetch()
+        for i in range(20):                       # check max 20 reads
+            try:
+                bam_handler.mate(next(bam_iter))  # Raises ValueError – if the read is unpaired or the mate is unmapped
+            except ValueError:
+                return False
+        return True
+
+
+def get_default_chr(args):
+    with pysam.AlignmentFile(args.bam, mode="rb", threads=args.threads) as bam_handler:
+        return [_.contig for _ in bam_handler.get_index_statistics()]
+
+
+def mate_mapped_to_intron(pileups):  # TODO
+    return pileups
+
+
+def count_alignments(column_iter, filter_by_mate=None):
+    filter_by_mate = False if filter_by_mate is None else filter_by_mate
+    collected_alignments = []
+    for column in column_iter:
+        if filter_by_mate:
+            collected_alignments.extend(mate_mapped_to_intron(column.pileups))
+        else:
+            collected_alignments.extend(column.pileups)
+    return len(collected_alignments)
+
+
+def process_contig(args, job):
+    contig = job[0]
+    location = job[1]
+    print(f"""Process {contig} into {location}""")
+    with open(location + "__" + contig, "w") as output_stream:
+        with pysam.AlignmentFile(args.bam, mode="rb", threads=args.threads) as bam_handler:
+            with pysam.TabixFile(args.ref, mode="r", parser=pysam.asBed(), threads=args.threads) as ref_handler:
+                for intron in ref_handler.fetch(contig):
+                    five_prime_counts = count_alignments(
+                        column_iter=bam_handler.pileup(       # should return exactly one item
+                            contig=contig,
+                            start=intron.start-args.span,
+                            end=intron.start-args.span+1,

```

Comment:
This doesn't do what I think it should, though it could certainly be my interpretation that's wrong.

Let's say you have an exon (=)/intron (-) junction that looks like this:
```
===========-------------
```
With a span of 3, I would expect to be looking for reads in a window 3bp on either side of the junction:

```
===========-------------
        ******
```

or possibly a window of 3bp at the start of the intron (not sure I like that interpretation though):

```
===========-------------
           ***
```

this code looks at a window 1bp wide that is 3bp upstream of the junction:

```
===========-------------
        *
```

If my interpretation is correct, this pileup should be:
```
start=intron.start-args.span,
end=intron.start+args.span,
```

with no `+1` needed since it's bracketing a zero-width event (the exon/intron junction), not a specific nucleotide.  Same logic would apply to the 3' counts below.

Also, I assume (but haven't checked) that the introns defined in `args.ref` are listed in genomic coordinates, so that `start` is always a lower coordinate than `end`, no matter which strand the transcript is on